Dockerfile

```diff
@@ -1 +1,41 @@
+ARG ARCH="amd64"
+ARG OS="linux"
+FROM 345280441424.dkr.ecr.ap-south-1.amazonaws.com/ark_base:latest
+LABEL maintainer="Armedia, LLC"
 
+ARG ARCH="amd64"
+ARG OS="linux"
+ARG VER="2.28.1"
+ARG PKG="prometheus"
+ARG SRC="${PKG}-${VER}.${OS}-${ARCH}"
+
+RUN curl \
+		-L "https://github.com/prometheus/${PKG}/releases/download/v${VER}/${SRC}.tar.gz" \
+		-o package.tar.gz && \
+	tar -xzvf "package.tar.gz" && \
+	mkdir -pv "/prometheus" && \
+	mkdir -pv "/etc/prometheus" && \
+	mkdir -pv "/usr/share/prometheus" && \
+	mv -vif "${SRC}/LICENSE"                     "/LICENSE" && \
+	mv -vif "${SRC}/NOTICE"                      "/NOTICE" && \
+	mv -vif "${SRC}/prometheus"                  "/bin/prometheus" && \
+	mv -vif "${SRC}/promtool"                    "/bin/promtool" && \
+	mv -vif "${SRC}/prometheus.yml"              "/etc/prometheus/prometheus.yml" && \
+	mv -vif "${SRC}/console_libraries/"          "/usr/share/prometheus/" && \
+	mv -vif "${SRC}/consoles/"                   "/usr/share/prometheus/" && \
+	ln -s "/usr/share/prometheus/console_libraries" "/etc/prometheus" && \
+	ln -s "/usr/share/prometheus/consoles"          "/etc/prometheus" && \
+	chown -R nobody:nobody "/etc/prometheus" "/prometheus" && \
+	rm -rvf "${SRC}" package.tar.gz
+
+#COPY npm_licenses.tar.bz2                        /npm_licenses.tar.bz2
+
+USER       nobody
+EXPOSE     9090
+VOLUME     [ "/prometheus" ]
+WORKDIR    /prometheus
+ENTRYPOINT [ "/bin/prometheus" ]
```

Comment:
Please do not override the `ENTRYPOINT`. There must be a very good reason to do that and it usually is a bad idea in 99% of the cases.

Reply:
I'm confused ... nearly every Dockerfile I've seen uses ENTRYPOINT vs. CMD ... are you saying that you prefer the use of CMD to ENTRYPOINT?

Reply:
`CMD` is totally the preferred way to specify what you want to run inside your Docker container. I my opinion, there are only 2 valid reasons why you would want to override `ENTRYPOINT`:

1. You want to provide the user with a convenient way to use the same Docker image for two different purposes. For example, I think the official MariaDB image can run either as client or server, and they override `ENTRYPOINT` to achieve this.
2. You absolutely need to run some script when your Docker image is started, even if you run `docker run -it MYIMAGE sh`.

By default, `ENTRYPOINT` is set to `/bin/sh -c`, which allows to run anything. If you set it to, say, `/bin/prometheus`, then you won't be able to run, say, `docker run -it MYIMAGE sh`, because `sh` is not an argument for `/bin/prometheus`; you can still manage to run `docker run` by overriding the overriden `ENTRYPOINT` and set it back to `/bin/sh -c`.

Reply:
Fair enough. I'll adjust as needed. I don't think I've seen this issue, though - I'm fully able to launch shells on images which have an ENTRYPOINT declared.

As I read the docs, ENTRYPOINT is meant to point at the executable which is to be run by the container, and CMD is meant to give the default parameters to that executable.

Reply:
Ok so here's a concrete example of why using CMD as the base vs. ENTRYPOINT is a bad idea ... take this docker-compose.yml fragment:

`
version: "3"

services:
    prometheus:
        image: backups.rivera.prv:6088/ark_prometheus:latest
        container_name: prometheus
        dns_search: rivera.prv
        dns:
            - 192.168.3.253
            - 192.168.3.252
        volumes:
            - ./data/prometheus:/app/data
            - ./conf/prometheus:/app/conf
        ports:
            - 9090:9090
        depends_on:
            - cadvisor
            - pushgateway
        restart: unless-stopped
        command:
            - '--config.file=/app/conf/prometheus.yml'
            - '--storage.tsdb.path=/app/data'
            - '--web.console.libraries=/usr/share/prometheus/console_libraries'
            - '--web.console.templates=/usr/share/prometheus/consoles'
`

When using CMD, the command would also have to include the executable one wishes to run. That kind of defeats the purpose of having *application containers*. With ENTRYPOINT, the above works just fine b/c it assumes the base command that will be used, and thus one can limit oneself to tuning parameters as appropriate.

This also means that a mistake on a developer's part can lead to the wrong executable being launched on an image ...

I'm going to keep ENTRYPOINT until we figure this out. Using CMD in the manner suggested is, in my mind, a BAD idea. I hadn't found a reason to oppose it, until right this second when I re-tried to rebuild the actual stack these images will be a part of.

Hopefully you guys have a solution for the above scenario and we can all come to a happy convergence :).

Reply:
@drivera-armedia Sure, I see your point... I see pros and cons for both aspects. I am kind of changing my mind on this and I tend to think you are right.